### DISCUSSION TITLE
How should the project player be themed?

### DISCUSSION THREAD

**Discussion by lisa-wolfgang** [ScratchAddons/ScratchAddons#4863] [Discussion]
Following the reasoning of https://github.com/ScratchAddons/ScratchAddons/issues/3014, it doesn't make sense that a theme for the editor should affect something that appears in multiple locations. Currently, on the scratch.mit.edu domain, there are a total of **five locations** where the project player can appear:
1. On the project page
2. In the editor
3. On the full screen page
4. On the embed page (minor factor)
5. On profile pages (when the live featured project addon is enabled) -- and by the way, this addon currently doesn't respect either theme addon

In my opinion, it makes little sense to theme every appearance of the project player after the editor when it has an arguably more prominent appearance in a non-editing context. Here are some options to consider:
- Give the project player its own theme addon
- Theme the project player dynamically based on the context (full screen is themed the same as the page it was initialized from)
- Make the project player follow the website theme (not preferable)

**Comment by Samq64** [comment]
I like the first one, but would be tricky or impossible to do with just css, so I think the second one is the best.

**Reply by Secret-chest** [reply]
We can use some JS to chexk what page one came from and apply the styles

**Reply by Samq64** [reply]
But the styles would have to be applied with JS and for thatb many the code would be pretty messy.

**Reply by lisa-wolfgang** [reply]
Many addons bundle "programmatic CSS" under a single class in a userstyle. Then, only the class needs to be toggled to toggle all of the styles.

**Reply by mxmou** [reply]
We already have `.sa-body-editor`.

**Reply by Samq64** [reply]
How does that work?

**Reply by mxmou** [reply]
The content script adds the `sa-body-editor` class to the body element in the editor and in full screen mode.

**Comment by mxmou** [comment]
Another option is to add a separate addon but use the editor/website dark mode settings if it's disabled.

> (full screen is themed the same as the page it was initialized from)

There is currently (almost, because of LLK/scratch-www#4190) no visible difference between the editor and project page full screen mode and I don't think we should change that. If we choose this approach, we should choose one addon that styles the full screen view.

**Reply by WorldLanguages** [reply]
I don't think it makes sense to style the fullscreen differently depending how it was opened.

**Reply by WorldLanguages** [reply]
> Another option is to add a separate addon but use the editor/website dark mode settings if it's disabled.

I think this could work. But I'm not sure if the complexity a third addon would bring is worth it.

**Comment by lisa-wolfgang** [comment]
Based on the discussion so far, I think the best course of action would be to create a separate theme for the project player and treat full screen mode as an extension of the project player -- so we would provide options for things like full screen background color, for example, in this new addon. This is the most logical connection because full screen mode is always directly accessed through the project player, and for this reason, I would assume that most users (with little technical knowledge about Scratch) also think of full screen mode as an extension of the project player. It would be best if we also [moved the project player addons to a top-level category](https://github.com/ScratchAddons/ScratchAddons/pull/3955) in order to increase this theme's visibility.

Theming when the addon is disabled is a little trickier. Implementation complexity and how much sense the fallback(s) make both need to be considered here. Since the project player is logically an extension of the project player, the most logical fallback would be the context that the project player is in. This would require some additional style overrides for the project page, but it would otherwise not require much additional effort. For full screen mode, if we're following the same logic as when the addon is enabled, it should be themed after the context that the project player was opened from.

**Reply by WorldLanguages** [reply]
I think we could create an issue about having a dedicated project player theme, and move the discussion there. Or does it exist already? I don't think so

**Comment by Hans5958** [comment]
At that point we should make a themes system for all of Scratch (#2232)
